File: ip_monitor.pl
#!/usr/bin/perl

use strict;
use YAML::XS qw(LoadFile DumpFile);
use HTTP::Tiny;
use Email::Send;
use Email::Send::Gmail;
use Email::Simple::Creator;

my $config_file  = "ip_monitor.yml";
my $log_filename = "ip_monitor.log";

open(my $log_fh, '>>', $log_filename) or die "Cannot open log file ($log_filename) for appending.\n$!\n";

my $config = LoadFile($config_file);

sub output {
    my $action = shift;
    my $message = shift;
    my $out = sprintf("%04d-%02d-%02d %02d:%02d:%02d: %s\n", (localtime)[5]+1900, (localtime)[4]+1, (localtime)[3,2,1,0], sprintf($message, @_));

    print $log_fh $out;

    if($action eq 'email' or $action eq 'warn' or $action eq 'die') {
        my $email = Email::Simple->create(
            header => [
                From    => $config->{config}->{email_from},
                To      => $config->{config}->{email_to},
                Subject => sprintf("IP Address Monitor"),
                ],
            body => sprintf("%s", $out)
            );

        my $sender = Email::Send->new(
            {   mailer      => 'Gmail',
                mailer_args => [
                    username => $config->{config}->{username},
                    password => $config->{config}->{password},
                ]
            }
        );
        $sender->send($email) or warn "Error sending email: $@";
    }

    if($action eq 'warn') { warn $out; }
    elsif($action eq 'die') { die $out; }
    else { print $out; }

    return;
}

my $url = $config->{api_url};

my $response = HTTP::Tiny->new->get($url);
output('warn', 'IP API call failed.') unless $response->{success};

my $new_address = $response->{content}; 

my $last_address = $config->{last_address};


unless ( $last_address eq $new_address ) {
    output('email', "The Public IP Address has changed from %s to %s.", $last_address, $new_address);
    $config->{last_address} = $new_address;
    DumpFile($config_file, $config);
} else {
    output('print', "Public IP unchanged: %s", $new_address);
}

File: ip_monitor.py
#!/usr/bin/python3

import logging
import os
import requests
import smtplib
import socket
import ssl
import yaml

CONFIG_FILENAME = "ip_monitor.yml"

config = []
with open(CONFIG_FILENAME, 'r') as stream:
    try:
        config = yaml.safe_load(stream)
    except yaml.YAMLError as exc:
        print(exc)

logging.basicConfig(filename=config['config']['log_filename'], encoding='utf-8', level=logging.INFO, format='%(asctime)s %(message)s', datefmt='%m/%d/%Y %H:%M:%S')

def email(message):
    email_text = f"""\
From: {config['config']['email_from']}
To: {config['config']['email_to']}
Subject: IP Address Monitor

{message}
"""

    try:
        server = smtplib.SMTP('smtp.gmail.com', 587)
        server.starttls()
        server.login(config['config']['username'], config['config']['password'])
        server.sendmail(config['config']['email_from'], config['config']['email_to'], email_text)
        server.close()

        logging.warning('Email sent to %s.', config['config']['email_to'])
    except Exception as e:
        logging.error('Something went wrong while sending the failure email.')
        print(e)


result = requests.get(config['config']['api_url'])

if result.status_code != requests.codes.ok:
    message = f'IP API call failed. {result.text}'
    logging.error(message)
    email(message)
    config['last_address'] = 'error'
elif result.text != config['last_address']:
    message = f'The Public IP Address has changed from {config["last_address"]} to {result.text}.'
    logging.info(message)
    email(message)
    config['last_address'] = result.text
else:
    logging.info(f'Public IP unchanged: {result.text}')

with open(CONFIG_FILENAME, 'w') as stream:
    yaml.dump(config, stream, default_flow_style=False) 